src/common/state/redis.c

```diff
@@ -199,27 +263,102 @@ task *parse_redis_task_table_entry(task_id id,
-  REDIS_CALLBACK_HEADER(db, callback_data, r);
-
-  if (callback_data->done_callback) {
-    task_table_done_callback done_callback = callback_data->done_callback;
-    done_callback(callback_data->id, callback_data->user_context);
+  LOG_DEBUG("Calling object table add callback");
+  REDIS_MULTI_CALLBACK_HEADER(db, callback_data, r, requests_info);
+  redisReply *reply = r;
+  object_id id = callback_data->id;
+  unsigned char *digest = callback_data->data;
+
+#define NUM_CHECK_AND_SET_COMMANDS 3
+#define CHECK_AND_SET_SETNX_INDEX 0
+#define CHECK_AND_SET_GET_INDEX 1
+#define CHECK_AND_SET_SADD_INDEX 2
+
+  /* Check that we're at a valid command index. */
+  int request_index = requests_info->request_index;
+  LOG_DEBUG("Object table add request index is %d", request_index);
+  CHECK(request_index <= NUM_CHECK_AND_SET_COMMANDS);
+  /* If we're on a valid command index, execute the current command and
+   * register a callback that will execute the next command by incrementing the
+   * request_index. */
+  int status = REDIS_OK;
+  ++requests_info->request_index;
+  if (request_index == CHECK_AND_SET_SETNX_INDEX) {
+    /* Atomically set the object hash and get the previous value to compare to
+     * our hash, if a previous value existed. */
+    requests_info->is_redis_reply = true;
+    status =
+        redisAsyncCommand(db->context, redis_object_table_add_callback,
+                          (void *) requests_info, "SETNX objhash:%b %b", id.id,
+                          sizeof(object_id), digest, (size_t) DIGEST_SIZE);
+  } else if (request_index == CHECK_AND_SET_GET_INDEX) {
+    /* If there was an object hash in the table previously, check that it's
+     * equal to ours. */
+    CHECKM(reply->type == REDIS_REPLY_INTEGER,
+           "Expected Redis integer, received type %d %s", reply->type,
+           reply->str);
+    CHECKM(reply->integer == 0 || reply->integer == 1,
+           "Expected 0 or 1 from REDIS, received %lld", reply->integer);
+    if (reply->integer == 1) {
+      requests_info->is_redis_reply = false;
+      redis_object_table_add_callback(c, reply, (void *) requests_info);
+    } else {
+      requests_info->is_redis_reply = true;
+      status = redisAsyncCommand(db->context, redis_object_table_add_callback,
+                                 (void *) requests_info, "GET objhash:%b",
+                                 id.id, sizeof(object_id));
+    }
+  } else if (request_index == CHECK_AND_SET_SADD_INDEX) {
+    if (requests_info->is_redis_reply) {
+      CHECKM(reply->type == REDIS_REPLY_STRING,
+             "Expected Redis string, received type %d %s", reply->type,
+             reply->str);
+      DCHECK(reply->len == DIGEST_SIZE);
+      if (memcmp(digest, reply->str, reply->len) != 0) {
+        /* If our object hash doesn't match the one recorded in the table,
+         * report the error back to the user and exit immediately. */
+        LOG_FATAL("Object hash collision while adding manager");
```

Comment:
Not a hash collision, but rather a hash mismatch. This is a pretty important error message.. maybe something like "Multiple different objects exist with the same object ID. This can arise when a nondeterministic task is executed multiple times, for example to reconstruct a missing object or speculatively."